## ***Sprint 7 Project: Streamlit.***

#### **_Stage 1: Importing Libraries._**

In [1]:
# ======================================================================
# THIRD-PARTY IMPORTS
# ======================================================================

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import streamlit as st
import plotly.graph_objects as go

# ======================================================================
# GLOBAL CONSTANTS
# ======================================================================

state = 8866

# ======================================================================
# IMPORT VALIDATION
# ======================================================================

print("Libraries imported and global configurations set.")

Libraries imported and global configurations set.


#### **_Stage 2: DataFrame Loading Process._**

In [2]:
# ======================================================================
# DATAFRAME LOAD.
# ======================================================================

df = pd.read_csv("C:\\proyectos_julian\\proyecto_sprint7\\vehiculos_f1\data\\vehicles_us.csv")

#### **_Stage 3: General Description of the Dataset._**

In [3]:
# ======================================================================
# GENERAL INFORMACTION OF THE DATAFRAME
# ======================================================================
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


#### **_Stage 4: Preliminary Visualization and Statistical Summary._**

In [4]:
# ======================================================================
# SHOW THE FIRST 5 ROWS.
# ======================================================================
df.head()

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
0,9400,2011.0,bmw x5,good,6.0,gas,145000.0,automatic,SUV,NaN,1.0,2018-06-23,19
1,25500,NaN,ford f-150,good,6.0,gas,88705.0,automatic,pickup,white,1.0,2018-10-19,50
2,5500,2013.0,hyundai sonata,like new,4.0,gas,110000.0,automatic,sedan,red,NaN,2019-02-07,79
3,1500,2003.0,ford f-150,fair,8.0,gas,NaN,automatic,pickup,NaN,NaN,2019-03-22,9
4,14900,2017.0,chrysler 200,excellent,4.0,gas,80903.0,automatic,sedan,black,NaN,2019-04-02,28


In [5]:
# ======================================================================
# SHOW 10 RANDOM ROWS.
# ======================================================================
df.sample(10, random_state=state)

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
36533,6995,2007.0,toyota sienna,good,6.0,gas,140715.0,automatic,van,black,NaN,2018-07-11,31
6335,22100,2013.0,ford f150 supercrew cab xlt,good,6.0,gas,80854.0,other,pickup,white,1.0,2018-08-31,31
33905,18995,2014.0,chevrolet tahoe,excellent,8.0,gas,110817.0,automatic,SUV,white,NaN,2018-08-21,30
40147,11900,NaN,gmc yukon,excellent,8.0,gas,165000.0,automatic,SUV,black,1.0,2018-06-09,16
51502,7995,2011.0,chevrolet equinox,like new,4.0,gas,111088.0,automatic,SUV,NaN,NaN,2019-04-01,175
8443,13900,2002.0,jeep wrangler,excellent,6.0,gas,NaN,automatic,SUV,blue,1.0,2018-09-07,36
1858,3500,1998.0,chevrolet camaro,good,6.0,gas,131528.0,automatic,coupe,blue,NaN,2018-11-28,12
7620,2695,1999.0,chevrolet malibu,good,6.0,gas,141000.0,automatic,sedan,green,NaN,2019-04-15,83
49166,33500,2017.0,ram 2500,excellent,6.0,diesel,137000.0,automatic,pickup,white,1.0,2019-01-27,41
48021,2000,2004.0,subaru outback,good,4.0,gas,NaN,automatic,wagon,white,1.0,2018-06-24,79


In [6]:
# ======================================================================
# SHOW THE LAST 5 ROWS.
# ======================================================================
df.tail()

,price,model_year,model,condition,cylinders,fuel,odometer,transmission,type,paint_color,is_4wd,date_posted,days_listed
51520,9249,2013.0,nissan maxima,like new,6.0,gas,88136.0,automatic,sedan,black,NaN,2018-10-03,37
51521,2700,2002.0,honda civic,salvage,4.0,gas,181500.0,automatic,sedan,white,NaN,2018-11-14,22
51522,3950,2009.0,hyundai sonata,excellent,4.0,gas,128000.0,automatic,sedan,blue,NaN,2018-11-15,32
51523,7455,2013.0,toyota corolla,good,4.0,gas,139573.0,automatic,sedan,black,NaN,2018-07-02,71
51524,6300,2014.0,nissan altima,good,4.0,gas,NaN,automatic,sedan,NaN,NaN,2018-06-05,10


In [7]:
# ======================================================================
# SHOW STATISTICAL SUMMARY.
# ======================================================================
df.describe()

,price,model_year,cylinders,odometer,is_4wd,days_listed
count,51525.000000,47906.000000,46265.000000,43633.000000,25572.0,51525.00000
mean,12132.464920,2009.750470,6.125235,115553.461738,1.0,39.55476
std,10040.803015,6.282065,1.660360,65094.611341,0.0,28.20427
min,1.000000,1908.000000,3.000000,0.000000,1.0,0.00000
25%,5000.000000,2006.000000,4.000000,70000.000000,1.0,19.00000
50%,9000.000000,2011.000000,6.000000,113000.000000,1.0,33.00000
75%,16839.000000,2014.000000,8.000000,155000.000000,1.0,53.00000
max,375000.000000,2019.000000,12.000000,990000.000000,1.0,271.00000


##### **_Resumen general, Stage 4: Preliminary Visualization and Statistical Summary._**
* Utilizando .head() podemos visualizar las primeras 5 filas, lo que nos permite confirmar la correcta carga del DataFrame.
* Con .sample() se extrajeron filas aleatorias, lo que nos puede visualizar valores distribuidos en el cuerpo del DataFrame, asi mismo poder encontar valores atípicos.
* Con ayuda de .tail() nos muestra las ultimas 5 filas, verificando las ultimas filas del DataFrame, asi mismo podemos ver la consistencia del DataFrame hasta el final.
* Con .describe() generamos un resumen estadistico de las variables numericas, lo que facilita la comprensión inicla del los datos.
* Visulaizando las filas del DataFrame, podemos ver la presencia de valores ausenten, de los cuales tendremos que trabajar con ellos en el siguiente paso.

#### **_Step 5: Data Preparation._**

##### **_5.1 Column treatment._**
* Es importante verificar que los datos que se encuentran almaenados en cada una de las columnas obtengan un tratamiendo para poder trabajar con ellos mas adelante tal y como se hara a continuacion.

##### **_5.2 Check for missing values._**

In [8]:
# ======================================================================
# SHOW COLUMN "price".
# ======================================================================
df.isna().sum()

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

* Podemos notar presencia de valores ausentes en las columnas "model_year", "cylinders", "odometer", "paint_color" y "is_4wd".
* Se realizara una limpieza de las columnas que presentas estas columnas en especifico.

##### **_5.3 Column: price._**

In [9]:
# ======================================================================
# SHOW COLUMN "price".
# ======================================================================
print(df['price'])

0         9400
1        25500
2         5500
3         1500
4        14900
         ...  
51520     9249
51521     2700
51522     3950
51523     7455
51524     6300
Name: price, Length: 51525, dtype: int64


* La columna "price" almacena datos de manera segura y eficiente valores enteros, lo que nos lleva a una conclusion y es que no es necesario manejar algun tipo de conversion, ya que con estos datos se pueden trabajar de manera segura.

##### **_5.4 Column: model_year._**

In [10]:
# ======================================================================
# SHOW COLUMN "model_year".
# ======================================================================
df['model_year']

0        2011.0
1           NaN
2        2013.0
3        2003.0
4        2017.0
          ...  
51520    2013.0
51521    2002.0
51522    2009.0
51523    2013.0
51524    2014.0
Name: model_year, Length: 51525, dtype: float64

In [11]:
# ======================================================================
# AVERAGE OF NULL VALUES IN THE COLUMN: : "model_year".
# ======================================================================
promedio_model_year = df['model_year'].isnull().sum() / len(df["model_year"]) * 100
print(f'Promedio de valores nulos en la columna "model_year": {promedio_model_year:.2f}%')

# ======================================================================
# REMOVE MISSING VALUES OF THE COLUMN: "model_year".
# ======================================================================
df = df.dropna(subset=['model_year'])

# ======================================================================
# CONVERT THE VALUES TO INTEGERS OF THE COLUMN: "model_year".
# ======================================================================
df['model_year'] = df['model_year'].astype(int)

Promedio de valores nulos en la columna "model_year": 7.02%


* Se realizo el calculo promedio de los valores nulos de la columna "model_year" arrojandonos un resultado de: 7.02%, lo que representa una pequeña fraccion, lo que nos lleva a la conclucion de eliminar por completo esos valores nulos.

##### **_5.5 Column: model._**

In [12]:
# ======================================================================
# SHOW COLUMN "model".
# ======================================================================
df['model']

0                bmw x5
2        hyundai sonata
3            ford f-150
4          chrysler 200
5          chrysler 300
              ...      
51520     nissan maxima
51521       honda civic
51522    hyundai sonata
51523    toyota corolla
51524     nissan altima
Name: model, Length: 47906, dtype: object

* Podemos observar que la columna "model" maneja datos de tipo object, lo cual no es necesario realizar algun tipo de conversion, ya que estamos trabajando con una variable categorica.

##### **_5.6 Column: condition._**

In [13]:
# ======================================================================
# SHOW COLUMN "condition".
# ======================================================================
df['condition']

0             good
2         like new
3             fair
4        excellent
5        excellent
           ...    
51520     like new
51521      salvage
51522    excellent
51523         good
51524         good
Name: condition, Length: 47906, dtype: object

In [14]:
# ======================================================================
# SHOW UNIQUE VALUES OF THE COLUMN "condition".
# ======================================================================
df['condition'].nunique()

6

* En la columna "condition" podemos observar que maneja valores de tipo object y ademas podemos visualizar los valores unicos de la columna, no es necesario realizar alguna conversion acerca de los datos almacenados en esta columna.

##### **_5.7 Column: cylinders._**

In [15]:
# ======================================================================
# AVERAGE OF NULL VALUES IN THE COLUMN: : "cylinders".
# ======================================================================
promedio_cylinders = df['cylinders'].isnull().sum() / len(df["cylinders"]) * 100
print(f'Promedio de valores nulos en la columna: {promedio_cylinders:.2f}%')

# ======================================================================
# REMOVE MISSING VALUES OF THE COLUMN: "cylinders".
# ======================================================================
df = df.dropna(subset=['cylinders'])

Promedio de valores nulos en la columna: 10.22%


* Se tomo la desicion de eliminar los valores ausentes, dado que el promedio de los valores ausentes es de 10.22%, la base resultante del 89% de los registros son completamente originales, lo que nos puede llevar a tener un mejor enfoque en nuestros datos para trabajar mas adelante con ellos.

##### **_5.8 Column: fuel._**

In [16]:
# ======================================================================
# SHOW COLUMN "fuel".
# ======================================================================
df['fuel']

0        gas
2        gas
3        gas
4        gas
5        gas
        ... 
51520    gas
51521    gas
51522    gas
51523    gas
51524    gas
Name: fuel, Length: 43009, dtype: object

In [17]:
# ======================================================================
# SHOW UNIQUE VALUES OF THE COLUMN "fuel".
# ======================================================================
df['fuel'].unique()

array(['gas', 'diesel', 'hybrid', 'other', 'electric'], dtype=object)

* La columna "fuel" maneja datos de tipo object, por oyra parte podemos observar los valore unicos de la columna y podemos visualizar los valores unicos que maneja la columna, por lo que no es necesario realizar alguna modificacion extra en esta columna.

##### **_5.9 Column: odometer._**

In [18]:
# ======================================================================
# SHOW COLUMN "odometer".
# ======================================================================
df['odometer']

0        145000.0
2        110000.0
3             NaN
4         80903.0
5         57954.0
           ...   
51520     88136.0
51521    181500.0
51522    128000.0
51523    139573.0
51524         NaN
Name: odometer, Length: 43009, dtype: float64

In [19]:
# ======================================================================
# AVERAGE OF NULL VALUES IN THE COLUMN: : "odometer".
# ======================================================================
promedio_odometer = df['odometer'].isnull().sum() / len(df["odometer"]) * 100
print(f'Promedio de valores nulos en la columna: {promedio_odometer:.2f}%')

# ======================================================================
# REMOVE MISSING VALUES OF THE COLUMN: "cylinders".
# ======================================================================
df = df.dropna(subset=['odometer'])

# ======================================================================
# CONVERT THE VALUES TO INTEGERS OF THE COLUMN: "model_year".
# ======================================================================
df['odometer'] = df['odometer'].astype(int)

Promedio de valores nulos en la columna: 15.32%


* Se tomo la desicion de eliminar los valores ausentes de la columna ¨odometer¨. El promedio de los valores ausentes es de 15.32%, esta desicion fue llevada a cabo por el enfoque del DataFrame que se basa en los autos usados y no esta reflejado en los autos nuevos. Lo que nos da un enfoque mas especifico de como queremos trabajar con los valores ariginales que no fueron eliminados.

##### **_5.10 Column: transmission._**

In [20]:
# ======================================================================
# SHOW COLUMN "transmission".
# ======================================================================
df['transmission']

0        automatic
2        automatic
4        automatic
5        automatic
6        automatic
           ...    
51518    automatic
51520    automatic
51521    automatic
51522    automatic
51523    automatic
Name: transmission, Length: 36419, dtype: object

In [21]:
# ======================================================================
# SHOW UNIQUE VALUES OF THE COLUMN "fuel".
# ======================================================================
df['transmission'].unique()

array(['automatic', 'manual', 'other'], dtype=object)

* Podemos observar que la columna ¨transmission¨ maneja valores de tipo object, asi mismo podemos visualizar los valores unicos que almacena la columna, por lo que en conclucion podemos decir que no es necesario trabajar en la modificacion de estos datos.

##### **_5.11 Column: type._**

In [22]:
# ======================================================================
# SHOW COLUMN "type".
# ======================================================================
df['type']

0          SUV
2        sedan
4        sedan
5        sedan
6        sedan
         ...  
51518    sedan
51520    sedan
51521    sedan
51522    sedan
51523    sedan
Name: type, Length: 36419, dtype: object

In [23]:
# ======================================================================
# SHOW UNIQUE VALUES OF THE COLUMN "type".
# ======================================================================
df['type'].unique()

array(['SUV', 'sedan', 'pickup', 'truck', 'coupe', 'hatchback', 'van',
       'wagon', 'mini-van', 'convertible', 'other', 'bus', 'offroad'],
      dtype=object)

* Podemos observar que la columna ¨type¨ maneja valores de tipo object, ademas podemos visualizar que maneja valores unicos al aplicar .unique(), lo que nos lleva a la conclucion de no realizar algun cambio de los datos que se almacenan en esta columna.

paint color

##### **_5.12 Column: paint_color._**

In [24]:
# ======================================================================
# SHOW COLUMN "paint_color".
# ======================================================================
df['paint_color']

0           NaN
2           red
4         black
5         black
6         white
          ...  
51518    silver
51520     black
51521     white
51522      blue
51523     black
Name: paint_color, Length: 36419, dtype: object

In [25]:
# ======================================================================
# SHOW UNIQUE VALUES OF THE COLUMN "paint_color".
# ======================================================================
df['paint_color'].unique()

array([nan, 'red', 'black', 'white', 'grey', 'silver', 'custom', 'orange',
       'yellow', 'blue', 'brown', 'green', 'purple'], dtype=object)

In [26]:
# ======================================================================
# FILL MISSING VALUES WITH: UNKNOWN OF THE COLUMN "paint_color".
# ======================================================================
df['paint_color'] = df['paint_color'].fillna('unknown')
print(f'Valores nulos en la columna "paint_color": {df['paint_color'].isna().sum()}')

Valores nulos en la columna "paint_color": 0


* En la columna "paint_color" maneja valores ausentes referente al color de los autos, se tomo la desicion de rellenar esos valores ausentes con la palabra 'unknown' dando a expresar que el color del auto es desconocido, ademas de verificar los valores unicos que se manejan en la columna "paint_color" y no menos importante mencionar que los valores que se manejan en la columna son de tipo object, lo que no requiere alguna otra modificacion.

##### **_5.13 Column: is_4wd._**

In [27]:
# ======================================================================
# SHOW COLUMN "is_4wd".
# ======================================================================
df['is_4wd']

0        1.0
2        NaN
4        NaN
5        1.0
6        NaN
        ... 
51518    NaN
51520    NaN
51521    NaN
51522    NaN
51523    NaN
Name: is_4wd, Length: 36419, dtype: float64

In [28]:
# ======================================================================
# IMPUTE MISSING VALUES WITH 0 OF THE COLUMN "is_4wd".
# ======================================================================
df['is_4wd'] = df['is_4wd'].fillna(0)
print(f'Valores nulos en la columna "is_4wd": {df['is_4wd'].isna().sum()}')

# ======================================================================
# CONVERT THE VALUES TO INTEGER OF THE COLUMN "is_4wd".
# ======================================================================
df['is_4wd'] = df['is_4wd'].astype(int)

Valores nulos en la columna "is_4wd": 0


* Podemos verificar que los valores de la columna "is_4wd" manejan datos de tipo float, lo que se requiere manejar datos de tipo int, dicho procedimiento fue realizado con exito. tambien podemos notar presencia de valores ausentes, por lo que se dictamino que esos valores ausentes fueran inpotados con valor 0, esto con el fin de mantener la esencia del dataframe sin perder este tipo de valores, por el tratamiento de estos datos ahora quedaron en formato correcto.

##### **_5.14 Column: date_posted._**

In [29]:
# ======================================================================
# SHOW COLUMN "date_posted".
# ======================================================================
df['date_posted']

0        2018-06-23
2        2019-02-07
4        2019-04-02
5        2018-06-20
6        2018-12-27
            ...    
51518    2018-08-10
51520    2018-10-03
51521    2018-11-14
51522    2018-11-15
51523    2018-07-02
Name: date_posted, Length: 36419, dtype: object

In [31]:
# ======================================================================
# CONVERT TO PROPER FORMAT OH THE COLUMN "date_posted".
# ======================================================================
df['date_posted'] = pd.to_datetime(df['date_posted'])
#df.sample(5, random_state=state)

* Como podemos observar en la columna "date_posted" se encuentra en formato object, se tomo la decision de de convertir al formato correspondiente, en esta columna no hay valores nulos, por lo que no se requiere de otra conversion extra de los datos almacenados en la columna "date_posted".

##### **_5.15 Column: days_listed._**

In [32]:
# ======================================================================
# SHOW COLUMN "days_listed".
# ======================================================================
df['days_listed']

0        19
2        79
4        28
5        15
6        73
         ..
51518    63
51520    37
51521    22
51522    32
51523    71
Name: days_listed, Length: 36419, dtype: int64

* Podemos on=bservar que la columna "days_listed" no maneja valores nulos y los valores que se encuentran almacenados en la columna mantienen un formato de int64, los datos son correctos manejado a enteros por lo que no se requiere de alguna conversion extra en los datos almacenados en esta columna.

In [33]:
# ======================================================================
# We check the final changes.
# ======================================================================
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 36419 entries, 0 to 51523
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   price         36419 non-null  int64         
 1   model_year    36419 non-null  int64         
 2   model         36419 non-null  object        
 3   condition     36419 non-null  object        
 4   cylinders     36419 non-null  float64       
 5   fuel          36419 non-null  object        
 6   odometer      36419 non-null  int64         
 7   transmission  36419 non-null  object        
 8   type          36419 non-null  object        
 9   paint_color   36419 non-null  object        
 10  is_4wd        36419 non-null  int64         
 11  date_posted   36419 non-null  datetime64[ns]
 12  days_listed   36419 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(5), object(6)
memory usage: 3.9+ MB


* Podemos observar que las columnas del DataFrame ahora manejan valores adecuandos a su formato correcto.

In [36]:
# ======================================================================
# SHOW COLUMN "price".
# ======================================================================
df.isna().sum()

price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64

* Podemos notar que ya no hay valores ausentes, lo que nos lleva a concluir que la limpieza de datos fue todo un exito.